### ISSUE TITLE
Add revocation information to certificates

### ISSUE THREAD

**Issue by Yepoleb** [mitmproxy/mitmproxy#3140]
libcurl with the WinSSL/Schannel backend fails with an SSL error if no revocation information is provided in the forged certificate. This is preventing me from intercepting an application that uses a statically linked libcurl.

I was able to reproduce this issue using curl 7.60.0 configured with `--without-ssl --with-winssl`, compiled with MSYS2 mingw-w64-x86_64-gcc 7.3.0 on Windows 7. The full error message is `curl: (35) schannel: next InitializeSecurityContext failed: Unknown error (0x80092012) - The revocation function was unable to check revocation for the certificate.`

I've not yet figured out what exactly is required by curl to accept the certificate, but I'll update the issue once I do. Adding an unresolved URL might just be enough to satisfy it, considering the awful state of revocation in general.

##### System information

Mitmproxy: 3.0.3 
Python:    3.6.5
OpenSSL:   OpenSSL 1.1.0h  27 Mar 2018
Platform:  Linux-4.16.0-1-amd64-x86_64-with-debian-buster-sid

**Comment by Yepoleb** [comment]
I was able to make the revocation check pass by generating an empty CRL in DER format using OpenSSL and adding it to the certificate like this:

    cert.add_extensions(
        [OpenSSL.crypto.X509Extension(
            b"crlDistributionPoints", False, b"URI:http://mypc.lan/mitmproxy.crl")])

An unresolved URL was unfortunately not enough to pass, it needs to reachable.

**Comment by cortesi** [comment]
Thanks for reporting this. One possibility is to provide a CRL distribution point within mitmproxy under the magic "mitm.it" domain we use for cert installation. There are a couple of things to think about:

- It could cause privacy concerns because of the DNS lookup of mitm.it, which would go to our public DNS server by default, and occur even for unrelated traffic. 
- It could cause hard-to-diagnose issues in isolated networks where mitm.it doesn't resolve. 
- I'd want to be sure that including a CRL distribution point doesn't cause flow-on effects for other clients. We'd have to test quite carefully.

@mitmproxy/devs Thoughts?

**Comment by Yepoleb** [comment]
Thanks for your interest. I'm not sure what you mean by traffic going to your public DNS server. The initial domain query should never reach your servers, only the primary resolver which then anonymizes the request. I also can't think of a scenario where the certificate would get served but the CRL request still goes to the public mitm.it page.

**Comment by cortesi** [comment]
Right, we're on the same page. We should be able to serve the CRL request from within mitmproxy effectively. I've edited my response to clarify the point regarding DNS servers - I feel that DNS resolution for an unrelated domain whenever mitmproxy is used might be an issue, regardless of recursive resolution (it would be for me under some circumstances).

**Comment by Yepoleb** [comment]
I understand your privacy concerns. I think the best option would be to reuse the CRL URLs of the original cert but replace the path with a some hash that can easily be recognized by the proxy. This method would even improve stealth because those DNS lookups are no longer missing from potential log files. In my opinion this is as close to a perfect solution as possible and resolves all concerns apart from having to test carefully.

**Comment by mhils** [comment]
Can we maybe just do OCSP Stapling instead of dealing with CRLs?

**Comment by Yepoleb** [comment]
As this is purely a compatibility issue and has no impact on security, I don't think going with the option that offers the least compatibility but the best security is the best approach. Just stapling is probably not enough anyway as clients most likely don't check that unless a OCSP URI is present and then we need to operate a OCSP service too for all the ones that only support regular OCSP.

In my opinion stapling on its own is not an option, the decision can only be between regular OCSP (+ optional stapling), CRL or both.

**Comment by yoshimo** [comment]
would it be possible to implement an optional bogus crl link  first and later consider extending it with an active ocsp&crl responder?

**Comment by Yepoleb** [comment]
I've made a basic implementation in this fork: https://github.com/Yepoleb/mitmproxy-crl/tree/crl

It's very hacked together and does not quite match the concept previously discussed, but it's something I could give to people asking me for help :shrug:

**Comment by Yepoleb** [comment]
I improved the implementation in https://github.com/Yepoleb/mitmproxy-crl/commit/0a095b28cd15abe3bbc2da25563bc7899a6b3309 so it no longer uses a private domain and matches the behavior described in https://github.com/mitmproxy/mitmproxy/issues/3140#issuecomment-390504365. The concept can easily be extended to also provide OCSP URLs and responses. It's described in the commit message. But before further improving the implementation, is this something the developers would consider merging?

**Comment by yannickbt64** [comment]
Granted some work has been done regarding this, including code, and having myself problems regarding certificate revocation with libcurl going through mitmproxy on Windows 10... I wondered if these changes could be implemented in mainstream mitmproxy as it patches a hole in TLS handling for mitmproxy to appear even more "legit".

**Comment by yoshimo** [comment]
This is SIX years old and still not working with mitmproxy itself.
I agree with yannick.Making it closer to a real world CA would help a lot with convincing clients to accept it.

**Comment by mhils** [comment]
Do we have a definite answer to the question whether just OCSP stapling would be enough? I only see opinions but no facts further up in the thread.

If CRLs are indeed required, my preferred option would be to hardcode a CRL URL with an IP address that we control in all certs. This avoids any DNS leakage. We could then extend the onboarding app to also respond for that special IP.

**Comment by yoshimo** [comment]
How about an iterative approach to this?
An ocsp only patch for testing first and if that isn't enough add crl on top?

**Comment by yoshimo** [comment]
> Do we have a definite answer to the question whether just OCSP stapling would be enough? I only see opinions but no facts further up in the thread.

Facts can sometimes only be gathered by testing your code & assumptions in the wild.

> If CRLs are indeed required, my preferred option would be to hardcode a CRL URL with an IP address that we control in all 
I wouldn't do something completely static that allows apps to detect the mitm attempt. But again, this can be adjusted later if we find evidence that it is being abused.

**Comment by Yepoleb** [comment]
A OCSP stapling response or, as the standard calls it, Certificate Status Request can not be sent proactively, it has to be requested by the client. From [RFC 6066 section 8](https://datatracker.ietf.org/doc/html/rfc6066#section-8):

> a server MUST NOT send the "CertificateStatus" message unless it received a "status_request" extension in the client hello message

Many clients, including the mentioned curl with Schannel, do not request this by default or do not support it at all. Quote from the curl documentation on [CURLOPT_SSL_VERIFYSTATUS](https://curl.se/libcurl/c/CURLOPT_SSL_VERIFYSTATUS.html)

> This option works only with the following TLS backends: GnuTLS and OpenSSL

The issue of DNS leakage has already been solved in my fork and it's been proven to work. Hardcoding IPs is still a subtle information leak and may cause problems with transparent proxying, because traffic might be routed differently.

**Comment by yoshimo** [comment]
@mhils can't we just try with an early implementation like the one @Yepoleb  made in his fork and see how well it works?
This is one of the small things that other tools can do and earn money by that. Shouldn't be necessary.

**Comment by mhils** [comment]
> Many clients, including the mentioned curl with Schannel, do not request this by default or do not support it at all.

Thank you - this answers my question in https://github.com/mitmproxy/mitmproxy/issues/3140#issuecomment-2161632685: OCSP Stapling apparently would not fix compatibility. Apologies for missing your last reply.

If you have an approach that makes sense and you would like to upstream it, please open a PR. :)

**Comment by JordanPlayz158** [comment]
> If you have an approach that makes sense and you would like to upstream it, please open a PR. :)
Is Yepoleb's approach fine? As I got it working in latest mitmproxy (thankfully not much seemed to change in all this time), I could open a PR if Yepoleb's approach is fine (seems to be the case given `OCSP Stapling apparently would not fix compatibility.`)

**Comment by mhils** [comment]
From a cursory glance it looks great, I haven't looked too closely as there
is no pull request yet. :)

**Comment by JordanPlayz158** [comment]
Ah, that is fair, I will make one then

**Comment by Yepoleb** [comment]
Thanks for taking on the work of rebasing my changes! Please tag me when the PR is ready.

**Comment by JordanPlayz158** [comment]
No problem, believe me 99% of it was copy and paste, but yeah the code coverage and such may be a bit difficult (writing tests in python, and figuring out the testing framework, and mocking things and such, am looking at the other tests for reference)
